### PR TITLE
Freqs (derived from fastai) approach

### PR DESCRIPTION
**Pull request by Nozdi**

### CODE REVIEW

File: src/artificial_contrast/settings.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+SEED = 42
+
+TEST_SET_EXAMINATIONS = [
```
Comment (grechut) [inline src/artificial_contrast/settings.py:3]: That shouldn't be part of repository IMO.
Maybe source file `test_set.csv` ?

Or maybe even better => test set patients should reside in a fully separated directory.
Then you just provide different data path
```
data_train
    P1
    P2
data_test
    P112
```

File: src/artificial_contrast/data.py [hunk 2]
```diff
@@ -1,26 +1,38 @@
...
     for patient in patients:
         bc_dir = Path(data_path / patient / 'BC')
         scans += [str(p) for p in bc_dir.ls()]
 
     return sorted(scans)
 
 
-def get_patients(data_path):
-    return sorted(os.listdir(data_path))
+def get_patients(data_path, test=False):
```
Comment (grechut) [inline src/artificial_contrast/data.py:23]: IMO (as written later) `get_patients` shouldn't change with logic below, but extract test patients to `test_data_path`.
Asserts prolly belong to `scripts` but placing it here has a nice feature of "ensuring assert early". Maybe that expected counts should be just and optional keyword to pass

File: src/artificial_contrast/freqs.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def interp_1d(x: Tensor, xp: Tensor, fp: Tensor):
+    "Same as `np.interp`"
+    slopes = (fp[1:] - fp[:-1]) / (xp[1:] - xp[:-1])
+    incx = fp[:-1] - (slopes * xp[:-1])
+    locs = (x[:, None] >= xp[None, :]).long().sum(1) - 1
+    locs = locs.clamp(0, len(slopes) - 1)
+    return slopes[locs] * x + incx[locs]
+
```
Comment (grechut) [inline src/artificial_contrast/freqs.py:54]: Side-question/idea:
Maybe would be worth to have a test on a very simplified example

File: src/artificial_contrast/freqs.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+def open_dcm_img_factory(windows: List[float], freqs: Tensor) -> Callable:
+    def open_dcm_image(file_name, *args, **kwargs) -> Image:
+        min_val, max_val = windows
+        arr = hist_scaled_px(file_name, freqs, min_val, max_val)
```
Comment (grechut) [inline src/artificial_contrast/freqs.py]: Where is `hist_scaled_px` declared? I cannot see its declaration anywhere

File: scripts/get_freqs_bins.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+DATA_PATH = Path(os.environ['DATA'])
+FREQS = os.environ.get('FREQS', 'freqs.npy')
+WINDOWS = [-100, 300]
+
+
+def remove_outside_window(
+    tensor: torch.Tensor, windows: List[int] = WINDOWS
+) -> torch.Tensor:
+    min_val, max_val = windows
+    tensor = tensor[(tensor >= min_val) & (tensor <= max_val)]
```
Comment (grechut) [inline scripts/get_freqs_bins.py:23]: Interesting construction, always returns 1D ? :)

```
In [2]: a = torch.tensor([[1, 2, 3], [4, 5, 6]])                                

In [3]: a[(a >= 2) & (a <=5)]                                                   
Out[3]: tensor([2, 3, 4, 5])

In [4]: a                                                                       
Out[4]: 
tensor([[1, 2, 3],
        [4, 5, 6]])

In [5]: a = a.view(3, 2)                                                        

In [6]: a                                                                       
Out[6]: 
tensor([[1, 2],
        [3, 4],
        [5, 6]])

In [7]: a[(a >= 2) & (a <=5)]                                                   
Out[7]: tensor([2, 3, 4, 5])
```

File: scripts/get_freqs_bins.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    return tensor
+
+
+scans = get_scans(DATA_PATH, test=False)
+sample_dcms = torch.stack(
```
Comment (grechut) [inline scripts/get_freqs_bins.py]: It looks a bit combo stack + tuple + list, but I assume that's the simplest way to achieve desired structure

File: scripts/freqs_experiment.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+# Experiment
+scans = get_scans(data_path)
+patients = get_patients(data_path)
+
+validation_patients = patients[:12]
+
+print('Number of patients: ', len(patients))
+print('Validation patients: ', validation_patients)
+
+tfms = get_transforms(
```
Comment (grechut) [inline scripts/freqs_experiment.py:61]: Extra experiment with more eager transforms?

File: scripts/get_freqs_bins.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+freqs = freqhist_bins(sample_dcms)
+np.save(FREQS, freqs.numpy())
+
+scaled_samples = torch.stack(
+    tuple([hist_scaled_img(scan, freq_bins=freqs) for scan in scans])
+)
+# for normalization
+mean, std = scaled_samples.mean(), scaled_samples.std()
+print(mean, std)
```
Comment (grechut) [inline scripts/get_freqs_bins.py:45]: Maybe they should be saved and further read along with corresponding `freqs.npy` file
